### DISCUSSION TITLE
How can I customize Typst citation?

### DISCUSSION THREAD

**Discussion by fredguth** [quarto-dev/quarto-cli#10440] [Q&A]
### Description

I have a Typst template which uses citenotes in the margin (a bibliography entry in the margin and another in the end of the document).   I am trying to create an extension to use this template.  But I can't find in typst-template (nor in typst-show, biblio, definitions, etc) where can I change the generated `#cite(<Tufte2001>, form: "prose")` to `#citenote(<Tufte2001>)`

Is `cite` customizable?

**Comment by mcanouil** [comment]
Could you share a **small self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository? The goal is to make it as easy as possible for us to recreate your problem so that we can fix it: please help us help you! Thanks.

Did you look at:
- https://quarto.org/docs/authoring/citations.html
- https://quarto.org/docs/output-formats/typst.html#bibliography

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by fredguth** [reply]
I did read that documentation in [https://quarto.org/docs/authoring/citations.html] which was very informative. One of the problems is that I can't find where a markdown `@Typst2001` becomes Typst `#cite(<Typst2001>, form:"prose")`.

Pandoc generates:
Cite
```          [ Citation
              { citationId = "Tufte2001"
              , citationPrefix = []
              , citationSuffix = [ Str "p.30" ]
              , citationMode = AuthorInText
              , citationNoteNum = 1
              , citationHash = 0
              }
          ]
```
I imagine I have to handle this here.  How?  Lua filter?

**Reply by fredguth** [reply]
I added bellow the link to a mwe I created.

**Comment by gordonwoodhull** [comment]
If you haven't done so already, you should use `keep-typ:true` to see the generated Typst.

If you are using Quarto markdown, the generated Typst probably uses native Typst references and citations like `@ref[Supplement]`

https://typst.app/docs/reference/model/ref/
https://typst.app/docs/reference/model/cite/

I don't see `#cite` being called explicitly in a quick test, but again we need a reproducible example.

IIUC the Typst reference syntax converts automatically to `#cite` elements and I don't know if Typst allows that to be customized to use your `#citenote()` function instead.

I think you are probably on the right track with the show-rule you mention in #10446, albeit with the limitation you describe.

**Comment by fredguth** [comment]
I have created a Minimum (Non) Working Example:
https://github.com/fredguth/mwe-citation-quarto-typst

My main problem is that I receive at the Typst side a `content`.  While I want to receive `key`, `form` and `supplement`.

**Reply by gordonwoodhull** [reply]
I do see the `#cite`s in your example, thank you.

It looks like Pandoc generates the dedicated reference syntax `@ref[Supplement]` if the citation uses only a supplement, and `#cite(...)` if there are other arguments.

https://github.com/jgm/pandoc/blob/720aa874023ac281e6844a96b66e3d5225d0af7e/src/Text/Pandoc/Writers/Typst.hs#L511-L535

So I think for all practical purposes, the answer to your original question:

> Is `cite` customizable?

is no, not today anyway.

You would need to change Pandoc in order to change to a different `#cite` function, or do some kind of messy post-processing. (While watching out for ones that use the simple reference syntax.)

EDIT: this could probably be done with a Lua filter that transforms `Cite`s into raw Typst output before the Typst writer gets to it.

**Reply by fredguth** [reply]
I "solved" in this very ugly way.... all with Typst.
Notice I am assuming that user will pass the parameter `dy` as `..  dy.-1cm `

https://github.com/fredguth/mwe-citation-quarto-typst/commit/df391429d3e09a59b63ac0ffc214c2a69c070eea

**Reply by fredguth** [reply]
<img width="1089" alt="image" src="https://github.com/user-attachments/assets/494664ac-fb8c-415c-a0a9-7b43d7da4c43">

**Reply by fredguth** [reply]
Another problem is that there is a linebreak after the ref.  For example `... like [1, p.30] <\n> to move... `

**Reply by gordonwoodhull** [reply]
Ok, great! Glad you got that working. Interesting technique, to limit the show rule by `.where(form:"prose")` to stop it from recursing. 👀 

I made a demo of using Lua filters for the same purpose, which I'll paste below for completeness.

`margin-cites.lua`
```lua
function Cite(cite)
  return pandoc.Inlines({cite, pandoc.RawInline('typst', '#place(dy:-2em, dx:100%+1cm)[#cite(<' .. cite.citations[1].id .. '>, form: "full")]')})
  -- quarto.log.output(cite)
end
```

In `article.qmd` metadata:

```lua
filters:
  - margin-cites.lua
```

I ran into the line break problem as well. I think this is because `#place` resolves as a block element. Enclosing the `#place(...)` expression with `#box` does remove the line break, but it breaks the positioning of the margin note.

We are interested in adding similar margin layout features to Quarto's Typst format, in parity with our HTML and PDF formats:
https://quarto-dev.github.io/quarto-gallery/page-layout/tufte.html

So I'm interested to see where your extension goes!

There is also this Typst template to look at:
https://github.com/nogula/tufte-memo/tree/main

It could be made into a custom Typst format for Quarto, like these:
https://quarto.org/docs/output-formats/typst-custom.html

(This might need some Lua filter trickery, for the same problem that started this ticket, changing from `#cite` to something else.)

**Reply by fredguth** [reply]
Thanks for Lua filters!  I would take ages to do the same few lines of code of margin-cite.lua as I have never dealt with pandoc 
Funny fact:  I created a Tufte layout for Typst a year ago.  But the tufte-memo typst format is much better than mine and I have been using it as inspiration.

**Comment by fredguth** [comment]
I think with this change here it is closed:

https://github.com/fredguth/mwe-citation-quarto-typst

@gordonwoodhull Could you take a look?

**Reply by fredguth** [reply]
Tried to add the filter in _extension.yml but still needed in .qmd. Is it possible to include in the extension and do not require the user to include this in the frontmatter?

**Reply by mcanouil** [reply]
No, that's how filter works. You need to opt-in. Installing extensions does not do anything on its own.

The alternative is to define a format using the filter, thus you could use the format which would include the specific configuration.

**Reply by fredguth** [reply]
Is there an example of a format that uses specific filters?

**Reply by fredguth** [reply]
Just noticed there is an error in the code I sent. It is creating a linebreak after the citation.

**Reply by mcanouil** [reply]
I don't have one out of my mind, but it's exactly as if you set it in your document and following the documentation https://quarto.org/docs/extensions/formats.html

```
format:
  whatever:
    filters:
      - name.lua
```

**Comment by fredguth** [comment]
Noticed that the problem with linebreak is the reason to use `drafting`.
I changed the code to use drafting and it seems correct to me.
I couldn't change the margin settings in drafting, but that seems a problem with the package and I will consider this issue here closed.

https://github.com/fredguth/mwe-citation-quarto-typst

**Comment by fredguth** [comment]
In the same gist, can I customize the `html` citation? I want to remove the directive "dy.-3em" from the citation `locator`.
I also noticed that the CSL is not rendering locator in the bib-entry.  Is it something I can/should add with a lua filter or there is a way to set this in the CSL definition? I ask because I have seen CSLs that you write a locator like pp.30 and it changes to p.30 or vice-versa.

**Reply by mcanouil** [reply]
Please open a new discussion as HTML is not the same as Typst.
Also, please provide a small reproducible document when asking questions. It helps as the answers about customisation are never a simple option.

For instance, here an example would clarify what you want to modify and the context.
